File: machinePlayer.py
import numpy as np
import random


class Player():

    def __init__(self, row, column):
        self.table = np.zeros((row, column))
        self.path = []  # [[rest coin(s), take number], ...]
        self.learningRate = 0.5
        self.discountRate = 0.5

    def refreshPath(self):  # Refresh path only if the game is over.
        self.path.clear()

    def pathAppend(self, theRestCoin, take_i):
        self.path.append([theRestCoin, take_i])

    # Receive feedback either if the game is over or if insufficient coins to pick.
    def receiveFeedback(self, winOrLose):
        if winOrLose == 1:
            self.table[self.path[-1][0]-1][self.path[-1][1]-1] += 10
        elif winOrLose == -1:
            self.table[self.path[-1][0]-1][self.path[-1][1]-1] -= 10
        elif winOrLose == -2:  # receive this if an invalid move has been made
            lastPlace = self.path[-1][0]
            # a deduction(reverse) of the values that last invalid move added.
            self.givePoints(theRestCoin=lastPlace, reverse=True)
            self.table[lastPlace-1][self.path[-1][1]-1] -= float("inf")
            self.path.pop()
            return self.makeMove(lastPlace)

    ####################################################################################
    # Key Logic Here!
    def givePoints(self, theRestCoin, reverse=False):
        def calculatePoint(theRestCoin, take_i, choices):
            point = 1
            zero = np.array([0, 0, 0])
            if theRestCoin-1-(take_i+1) >= 0:
                target = self.table[theRestCoin-1-(take_i+1)::-1][:choices]
                for each in target:
                    # If there exists any row that all member in taht row are negative,
                    # it means that this choice will lead you to failure.
                    if (each < zero).all():
                        # point = -10   # How about sum up or choose minimum?
                        return target.min()*self.discountRate

                # If there exists any row that all member in taht row are zeros
                # (and meanwhile no row is all-negative,) it means that this choice
                # will lead you to an unknown area.
                for each in target:
                    if (each == zero).all():
                        point = 0
                        return point
                return target.max()*self.discountRate
            else:
                point = 0
                return point
        plusOrMinus = -1 if reverse else 1
        for i in range(1, 4):
            if i <= theRestCoin:
                # Q <- Q + lr*point
                self.table[theRestCoin-1][i-1] += self.learningRate * calculatePoint(
                    theRestCoin=theRestCoin, take_i=i, choices=3) * plusOrMinus
    ##################################################################################

    def makeMove(self, theRestCoin):
        # If an invalid move has been made (and has been discovered later on,)
        # a deduction of the values in each blank is needed.
        # But when the largest number you can choose is less than 4,
        # this kind of situation can actually be ignored.
        self.givePoints(theRestCoin)
        max_val = self.table[theRestCoin-1][0]
        take_i = 1
        thisRow = self.table[theRestCoin-1][:]
        if np.all(thisRow == thisRow[0]) and thisRow[0] == 0:
            # print(thisRow)
            take_i = random.randint(1, 3)
            # take_i = random.randint(1,min(3,theRestCoin))
        else:
            for i in range(1, 4):
                if thisRow[i-1] > max_val:
                    max_val = thisRow[i-1]
                    take_i = i
        self.pathAppend(theRestCoin=theRestCoin, take_i=take_i)
        return [take_i, max_val]

File: webpage/modules/machinePlayer.js
export class MachinePlayer {
    constructor(totalCoin, maxPickable) {
        this._totalCoin = totalCoin;
        this._maxPickable = maxPickable;
        this._table = Array(totalCoin).fill(null).map(() => Array(maxPickable).fill(0));
        this._path = []; // [[rest coin(s), number took], ...]
        this._currentPick = 0;
        this._lr = 0.5;
        this._discountRate = 0.5;
    }
    get currentPick() {
        return this._currentPick;
    }
    get table() {
        return this._table;
    }
    appendOnPath(coinRest, numTook) {
        this._path.push([coinRest, numTook]);
    }
    refreshPath() {
        this._path = [];
    }
    receiveFeedback(feedBack) {
        const targetRowIdx = this._path[this._path.length - 1][0] - 1;
        const targetColumnIdx = this._path[this._path.length - 1][1] - 1;
        if (feedBack == 1) { // win
            this._table[targetRowIdx][targetColumnIdx] += 10;
        }
        else if (feedBack == -1) { // lose
            this._table[targetRowIdx][targetColumnIdx] -= 10;
        }
        else if (feedBack == -2) { // if an invalid move has been made
            // make a deduction (reverse) of the values the latest invalid move added.
            // "targetRowIdx + 1" is the "latest number of coins rest"
            this.updateScoresOfARow(targetRowIdx + 1, true);
            this._table[targetRowIdx][targetColumnIdx] = -1 * Infinity;
            this._path.pop();
            this.makeMove(targetRowIdx + 1);
        }
    }
    makeMove(restCoin) {
        // learn something before make choice
        this.updateScoresOfARow(restCoin, false);
        const targetRow = this._table[restCoin - 1];
        let numTake;
        let maxScore;
        if (targetRow.every((i) => i == 0)) {
            numTake = Math.floor(Math.random() * this._maxPickable) + 1;
            maxScore = targetRow[numTake - 1];
        }
        else {
            maxScore = targetRow[0];
            numTake = 1;
            for (let i = 0; i < targetRow.length; i++) {
                if (targetRow[i] > maxScore) {
                    maxScore = targetRow[i];
                    numTake = i + 1;
                }
            }
        }
        this.appendOnPath(restCoin, numTake);
        this._currentPick = numTake;
    }
    updateScoresOfARow(restCoin, reverse) {
        const sign = reverse ? -1 : 1;
        for (let i = 1; i <= this._maxPickable; i++) {
            if (i <= restCoin) {
                // Q <- Q + lr * point
                this._table[restCoin - 1][i - 1] += sign * this._lr * this.calcScore(restCoin, i);
            }
        }
    }
    calcScore(restCoin, numTake) {
        const restCoinIfYouTakeN = restCoin - numTake;
        // If, afterwards, your opponent take "m" coins,
        // you will heve "restCoinIfYouTakeN - m" coins rest the next time.
        if (restCoinIfYouTakeN > 1) {
            let concernPart = [];
            for (let i = 0; i < Math.min(this._maxPickable, restCoinIfYouTakeN - 1); i++) {
                concernPart.push(this._table[(restCoinIfYouTakeN - 1) - 1 - i]);
            }
            // If there exists any row that all numbers are negative,
            // it means that "take i" now will lead you to failure
            // because your opponent is smart enough to push you to that situation.
            for (let eachRow of concernPart) {
                if (eachRow.every((i) => i < 0)) {
                    return Math.max(...eachRow) * this._discountRate;
                }
            }
            // If there exists any row that all numbers are zero
            // (and no row is all-negative) it means that this choice
            // will lead you to an unknown result.
            for (let eachRow of concernPart) {
                if (eachRow.every((i) => i == 0))
                    return 0;
            }
            // If non of the above two situations happened...
            let oneDArray = [];
            for (let eachRow of concernPart) {
                for (let eachScore of eachRow)
                    oneDArray.push(eachScore);
            }
            return Math.max(...oneDArray) * this._discountRate;
        }
        else {
            return 0;
        }
    }
}
